### DISCUSSION TITLE
Cognito + Third party IDPs - Nonce Mismatch Error

### DISCUSSION THREAD

**Discussion by leonp-s** [nextauthjs/next-auth#3551] [Help]
### Question 💬

I am currently trying to use AWS Cognito (setup with a third party IDP) however I am running into an issue when signing in with a third party IDP via Cognito. 

```
[next-auth][error][OAUTH_CALLBACK_ERROR]
https://next-auth.js.org/errors#oauth_callback_error nonce mismatch, expected undefined, got: ... {
  error: {
    message: 'nonce mismatch, expected undefined, got: ...',
    stack: 'RPError: nonce mismatch, expected undefined, got: ...
```

A similar question has been asked [here](https://github.com/nextauthjs/next-auth/issues/3403), and has been closed however I was hoping a solution might be proposed. 

### How to reproduce ☕️

1. Setup Cognito
2. Configure to Cognito user pool to use a third party IDP
3. Initiate authentication through IDP via Cognito

```
CognitoProvider({
      clientId: process.env.COGNITO_CLIENT_ID,
      clientSecret: process.env.COGNITO_CLIENT_SECRET,
      issuer: process.env.COGNITO_ISSUER,
    }),
```

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by dmcd** [comment]


**Reply by leonp-s** [reply]
Hi David, thanks for your reply. Just had another look at my scopes, they all seem to match across the user pool, provider config and IDP authorised scopes. I am using Google as a 3rd party IDP for Cognito requesting 'openid profile email'. Any other ideas?

**Reply by dmcd** [reply]
Apologies, the issue is unrelated to auth scopes. Its still happening for me when first completing the oauth flow. However, if I run the flow again, it completes ok and I get a valid session. I'm using cognito with apple sign in.

```
[next-auth][error][OAUTH_CALLBACK_ERROR]
https://next-auth.js.org/errors#oauth_callback_error nonce mismatch, expected undefined, got: t4csJhuaMJEk3YPYCu0XmQQDsic8qwcGaSK-3fYvsuE1qRORbrA15o47PFK-fO6yG5xHWEOAmpMtHopKwjUGiQmIvmALwYMAMEArd_7voN_9vA4m_Zt43Gtj6ICs8hSmM5tTymp5tkaOF6B6wWcSSS7xCuqnBeQpqRAUcX70uwo {
  error: {
    message: 'nonce mismatch, expected undefined, got: t4csJhuaMJEk3YPYCu0XmQQDsic8qwcGaSK-3fYvsuE1qRORbrA15o47PFK-fO6yG5xHWEOAmpMtHopKwjUGiQmIvmALwYMAMEArd_7voN_9vA4m_Zt43Gtj6ICs8hSmM5tTymp5tkaOF6B6wWcSSS7xCuqnBeQpqRAUcX70uwo',
    stack: 'RPError: nonce mismatch, expected undefined, got: t4csJhuaMJEk3YPYCu0XmQQDsic8qwcGaSK-3fYvsuE1qRORbrA15o47PFK-fO6yG5xHWEOAmpMtHopKwjUGiQmIvmALwYMAMEArd_7voN_9vA4m_Zt43Gtj6ICs8hSmM5tTymp5tkaOF6B6wWcSSS7xCuqnBeQpqRAUcX70uwo\n' +
      '    at Client.validateIdToken (src/node_modules/openid-client/lib/client.js:796:13)\n' +
      '    at processTicksAndRejections (internal/process/task_queues.js:93:5)\n' +
      '    at async Client.callback (src/node_modules/openid-client/lib/client.js:484:7)\n' +
      '    at async oAuthCallback (src/node_modules/next-auth/core/lib/oauth/callback.js:112:16)\n' +
      '    at async Object.callback (src/node_modules/next-auth/core/routes/callback.js:50:11)\n' +
      '    at async NextAuthHandler (src/node_modules/next-auth/core/index.js:133:28)\n' +
      '    at async NextAuthNextHandler (src/node_modules/next-auth/next/index.js:20:19)\n' +
      '    at async src/node_modules/next-auth/next/index.js:56:32\n' +
      '    at async Object.apiResolver (src/node_modules/next/dist/server/api-utils.js:102:9)\n' +
      '    at async DevServer.handleApiRequest (src/node_modules/next/dist/server/next-server.js:1064:9)',
    name: 'RPError'
  },
  providerId: 'cognito',
  message: 'nonce mismatch, expected undefined, got: t4csJhuaMJEk3YPYCu0XmQQDsic8qwcGaSK-3fYvsuE1qRORbrA15o47PFK-fO6yG5xHWEOAmpMtHopKwjUGiQmIvmALwYMAMEArd_7voN_9vA4m_Zt43Gtj6ICs8hSmM5tTymp5tkaOF6B6wWcSSS7xCuqnBeQpqRAUcX70uwo'
}
```

**Comment by rmelo** [comment]
I am also facing this problem, I already check my scopes and it's ok too. I'm also using Google and I noticed that in the first auth, when Google ask me for my credentials is when its happening, but if I sign out and sign in again, Google doesn't ask me because I already authenticated to it before, so the problem doesn't occur

**Comment by MaxOrelus** [comment]
Ditto. I've been running next-auth@4.0.0-beta.3 for months now and I had my Cognito setup all working. Today I decided to upgrade to 4.0.6 and I'm getting the same errors as the individuals above. ^^

Maybe a diff between 4.0.6 and the 4.0.0-beta.3 may pin point the issue.

**Comment by domwillcode** [comment]
I too am seeing this issue with Cognito. I've rolled back to 4.0.0-beta.3 for now as the latest functional build. 

4.0.0-beta-4 introduces an update to the 'openid-client' package from [^4.7.4](https://github.com/nextauthjs/next-auth/blob/v4.0.0-beta.3/package.json#L62) to [^5.0.1](https://github.com/nextauthjs/next-auth/blob/v4.0.0-beta.5/package.json#L67)  (although beta-4 is broken and fixed in beta-5) and now [^5.1.0](https://github.com/nextauthjs/next-auth/blob/v4.1.0/package.json#L73). It seems openid-client 5.0.1 forced id token validation (client.callback() instead of client.oauthCallback()) and the [nonce check in openid-client](https://github.com/panva/node-openid-client/blob/main/lib/client.js#L795). 

You can trigger the same validation by setting idToken true in your next-auth config for Cognito on 4.0.0-beta.3 (becoming the default from beta-5.

```
providers: [
            Cognito({
                idToken: true,
                ...
            )}
]
```

Cognito seems to generate the nonce for the initial auth when [using an IDP](https://docs.aws.amazon.com/cognito/latest/developerguide/amazon-cognito-user-pools-using-the-id-token.html) which triggers the error on callback. 

Beyond that, I'm stumped!

**Comment by lpsinger** [comment]
I'm getting exactly the same error message in my project which uses openid-client in a [Remix](https://remix.run) app. What I can't tell is if this is a bug in Cognito or a bug in openid-client.

**Comment by lpsinger** [comment]
Aha! I figured out a workaround for my code, which may work for next-auth too. Unfortunately my code isn't public (yet) so I can only provide a snippet here. In the call to `client.callback`, set `nonce` to `null` to skip the nonce check completely.

```js
  const tokenSet = await client.callback(getRedirectUri(request), params, {
    code_verifier,
    nonce: null  // <-- add this
  })
```

**Reply by lpsinger** [reply]
Here's a version that makes TypeScript happy, and includes some more comments.

```ts
  const tokenSet = await client.callback(getRedirectUri(request), params, {
    code_verifier,
    // @ts-expect-error: Set this to null to disable nonce check.
    //
    // When Cognito is federating a social IdP, it adds a nonce to the callback
    // request, which causes validation to fail, because we did not provide a
    // nonce.
    //
    // Setting nonce to null disables the nonce check entirely, but it angers
    // TypeScript because the nonce property is typed as string | undefined.
    nonce: null,
  })
```

**Reply by panva** [reply]
You've just disabled a very rudimentary check of openid connect. I would not call that a workaround, it's a dangerous action to say the least. Sounds like a bug in the IdP openid-client talks to, meaning Cognito.

**Reply by lpsinger** [reply]
Thanks @panva. I see you replied to the other discussion. Just repeating for the record on this discussion that I'll take this back to Cognito and report it to them as a bug.

**Reply by shairozan** [reply]
So everyone is aware, this is kind of a mixed bag. This _is_ something wrong with Cognito for one. See their below statement:

# AWS Issues

> In raising these questions to the Cognito Service Team, we also recognized that there was a documentation gap [1] when it came to mentioning nonces and how they interfaced with the authentication flow. For that we have raised a documentation update request that is already underway. Though I am unable to provide an ETA on that documentation update. 

> I understand that this is not the information you necessarily wanted to hear, as the nonce included in the ID token produced from your external SAML flow is breaking your auth flow designed around the OIDC spec---especially as the Cognito documentation did not include this information. I would be frustrated to find this information out after setting up an external auth flow for a user pool with SAML using the auth code grant, under the logical assumption that a nonce would not be generated for the auth code grant as it is listed as an optional parameter in the OIDC specifications [2]. 

> I apologize for any frustration caused by this lack of documentation and appreciate your patience with this process as I pursued escalation with the Cognito Service Team. I also would like to thank you for raising this case in the first place as this is now bringing about documentation detailing how nonces play into an external IdP flow for Cognito.

We noticed this and originally opened a ticket a month or so ago which was simply brushed off as "we don't support nonces". Ok cool, we're _not_ asking for a nonce, but a nonce is being generated in the ID token we get back. 

# Not only AWS

A secondary thing to note is that this is _only_ happening with Next Auth. We have internal tooling (similar to gimme-aws-creds) that operates on golang libs for Oauth2 / OIDC2 that does _not_ generate nonces in the ID token for the same:

* Authorization Type
* Client
* Scopes

It is easy to point out that Cognito is quite literally not adhering to the spec, but I'm also not seeing this behaviour in other OIDC / Oauth platforms if an upstream IDP is consumed through Cognito.

**Reply by lpsinger** [reply]
It's not just in Next Auth. I see this error message with node-openid-client (see panva/node-openid-client#456). I too contacted AWS support and they confirmed that this is a nonconformance of Cognito with the OIDC spec.

**Reply by james-bjss** [reply]
@shairozan - Sorry to bother you but we have run into this issue and are looking to raise/chase via AWS support (with the assumption this is an issue in the cognito service). Have you had any further updates or information back from AWS?

@panva  - I believe you have also raised this with AWS, have there been any useful updates from them?
https://github.com/panva/node-openid-client/discussions/456#discussioncomment-1978444

@lpsinger as noted by @panva your above solution of ignoring the nonce is not advised.

I noticed that you also had success by changing the auth url to provide an initial token. Do you know if there are any downsides to this workaround?

Assuming it's a viable workaround how are you achieving this is it by changing the `authorization` parameter in the `CognitoProvider` config to provide an initial nonce value?
https://github.com/panva/node-openid-client/discussions/456#discussioncomment-1986646

**Reply by james-bjss** [reply]
Digging into this a little more. It seems it doesn't help if we specify a nonce value in the URL in the  NextAuth Cognito provider as we can't set the `check` value that is used to compare the value returned.

The AWS [Cognito Documentation* ](https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-dg.pdf )has been updated to state that the nonce value is automatically set if none provided and it doesn't appear to be something on their radar to change.

Would a suitable solution/workaround be to provide a way to set or generate a nonce value in NextAuth's `CognitoProvider` and use that same value in the check? 

At the moment we can't set this :
See: https://github.com/nextauthjs/next-auth/blob/main/packages/next-auth/src/core/lib/oauth/callback.ts#L72

It's unfortunate as this is a total blocker for our team and we are now having to look at alternatives to NextAuth to build our Cognito integration.


*The Cognito docs have been updated to state:
```
Nonce (nonce)
The nonce claim comes from a parameter of the same name that you can add to requests to your
OAuth 2.0 authorize endpoint. When you add the parameter, the nonce claim is included in the
ID token that Amazon Cognito issues, and you can use it to guard against replay attacks. If you do
not provide a nonce value in your request, Amazon Cognito automatically generates and validates
a nonce when you authenticate through a third-party identity provider, then adds it as a nonce
claim to the ID token. The implementation of the nonce claim in Amazon Cognito is based on OIDC
standards.
```

**Reply by lpsinger** [reply]
> Digging into this a little more. It seems it doesn't help if we specify a nonce value in the URL in the NextAuth Cognito provider as we can't set the `check` value that is used to compare the value returned.

Can you store the check value in your session, the same as you would do with PKCE?

**Reply by james-bjss** [reply]
> > Digging into this a little more. It seems it doesn't help if we specify a nonce value in the URL in the NextAuth Cognito provider as we can't set the `check` value that is used to compare the value returned.
> 
> Can you store the check value in your session, the same as you would do with PKCE?

The issue is NextAuth provides no way to set the checks in the code (without modifying the library)
I don't think PKCE is supported for the Cognito provider ( i only see options under apple and others) but even then I think cognito is still returning a nonce value causing a mismatch error. 

Just digging into the code a bit further. Would be interesting to see what the Maintainers suggest as it seems Cognito is unusable with external IDPs at the moment.

**Edit:** I tweaked the provider to add the PKCE checks and saw its sending the correct params to cognito, but it still does the nonce check and fails.

**Reply by lpsinger** [reply]
> I don't think PKCE is supported for the Cognito provider

I am using PKCE with Cognito. (But I am using node-openid-client directly.)

**Reply by james-bjss** [reply]
> > I don't think PKCE is supported for the Cognito provider
> 
> I am using PKCE with Cognito. (But I am using node-openid-client directly.)

Yeah, I amended the provider to add the PKCE checks, but then NextAuth throws the nonce mismatch error after. Effectively it does this:

```js
    if (provider.token?.request) {
      // @ts-expect-error
      const response = await provider.token.request({
        provider,
        params,
        checks,
        client,
      })
      tokens = new TokenSet(response.tokens)
    } else if (provider.idToken) {
      tokens = await client.callback(provider.callbackUrl, params, checks)
    } else {
      tokens = await client.oauthCallback(provider.callbackUrl, params, checks)
    }
```

The checks value that gets passed to the open-id client can't be set externally and it never expects a nonce to be set in the reponse. ie. `const checks = {};` and the only value set in checks via the function relate to the state and PKCE checks. The openid-client expects a matching value in `checks.nonce`.

There is no means in the lib to provide the nonce value though, so if you add it to the auth url it sets it but then the check expects `undefined`. If you don't set one then Cognito (helpfully?) generates it's own and that also won't match `undefined`.

So assuming AWS won't budge I am wondering if we add the ability to generate or set this value when calling cognito and use this value in the check. Otherwise I don't see how anyone can use NextAuth with Cognito + and external IDP.

**Reply by balazsorban44** [reply]
I don't think we currently set or use `nonce` correctly, PRs are welcome!

**Reply by james-bjss** [reply]
Hi @balazsorban44 

I did consider raising a PR, the issue we have is if we define a `nonce` value in the provider config this value remains unchanged after it's initialized so assumed such a change would be rejected on security grounds (though currently no nonce is used at all so no worse off in some ways).

As a naive fix it would be easy to set the nonce value in the provider config,  then in the callback do something like:
`checks.nonce = options.provider.authorization.params.nonce` (obviously with null checks)

This could optionally behind a config in `protection` if we wanted to enable disable.

Where would you suggest that the value could be generated and stored as guessing defining it statically in the config is not ideal.

**Reply by james-bjss** [reply]
@balazsorban44  - We had a go at adding this functionality last night. Hopefully this looks reasonable. This has been added as an additional check that users can optionally enable in provider config and generates a new nonce value vs using a statically configured one.

https://github.com/nextauthjs/next-auth/pull/4100

**Comment by tiyberius** [comment]
Having the exact same issue as everyone here, errors out on first attempt at login, then works the second time. I'm new to Next-auth, really want to like it but this isn't a great first impression!

**Reply by james-bjss** [reply]
Hey @tiyberius we created a PR to resolve the issue. Not sure if it will be accepted by the maintainers though: https://github.com/nextauthjs/next-auth/pull/4100

**Reply by tiyberius** [reply]
Hey @james-bjss cheers for the reply and for the fix! I'm trying to test out your branch and I can't get anything going. I've tried

1. Cloning `git@github.com:james-bjss/next-auth.git`, checking out the `feature/nonce-check-type` branch, `yarn build`ing and then copying the `packages/next-auth` to my project's `node_modules`, but got the same callback error 🤔 
2. Installing the branch via yarn, by executing a `yarn add 'next-auth@https://github.com/james-bjss/next-auth.git#feature/nonce-check-type'` but getting
```
❯ yarn run dev
yarn run v1.22.17
$ next
ready - started server on 0.0.0.0:3000, url: http://localhost:3000
info  - Loaded env from /Users/tobias/tmp/next-auth-example/.env.local
error - ./pages/_app.tsx:1:0
Module not found: Can't resolve 'next-auth/react'
   > 1 | import { SessionProvider } from "next-auth/react"
     2 | import type { AppProps } from "next/app"
     3 | import "./styles.css"
     4 |
https://nextjs.org/docs/messages/module-not-found
```

Sorry for being a n00b, I'm relatively new to frontend dev!

**Reply by james-bjss** [reply]
Hey @tiyberius  we made enabling the nonce check optional.

In scenario 1 you just need to enable the check by including `nonce` string in the `checks` array in your provider config (see below). 
NB. You can also add `state` and `pkceCodeVerifier` to this array too. See the available checks here: https://github.com/nextauthjs/next-auth/blob/e65393356da273fcd53346d956f8c98bbbe6aa74/packages/next-auth/src/providers/oauth.ts

In scenario 2 the issue is that when you pull in the dependency you also need a `postinstall`  action to trans-pile from typescript to js. We need to set this up ourselves so will share our solution once we have done so.

Example enabling nonce check:

```js
import NextAuth from 'next-auth'
import CognitoProvider from 'next-auth/providers/cognito'

const handler = NextAuth({
  providers: [
    CognitoProvider({
      clientId: process.env.COGNITO_CLIENT_ID,
      clientSecret: process.env.COGNITO_CLIENT_SECRET,
      issuer: process.env.COGNITO_ISSUER,
      idToken: true,
      // Enable nonce validation check for cognito (can also add state and pkceCodeVerifier)
      checks: 'nonce',
    }),
  ],
})

export default handler
```

**Reply by tiyberius** [reply]
@james-bjss absolute legend!!! It works! Thanks heaps for the instructions 🙏🙏🙏 hopefully the maintainers can get this merged soon, it seems like a pretty basic functionality to be broken.

**Reply by james-bjss** [reply]
No worries. Glad it's useful for others!

**Reply by EvgenKoshevoy** [reply]
Hello. I still have same issue with Cognito Adapter (with SAML Azure AD). After clicking "Sign in with Cognito" again - i have access to page.

```
[next-auth][error][CALLBACK_OAUTH_ERROR]
https://next-auth.js.org/errors#callback_oauth_error nonce mismatch, expected undefined, got: Ys6...Oyw RPError: nonce mismatch, expected undefined, got: Ys6...Oyw
    at Client.validateIdToken (/app/node_modules/openid-client/lib/client.js:798:13)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async Client.callback (/app/node_modules/openid-client/lib/client.js:486:7)
    at async oAuthCallback (/app/node_modules/next-auth/core/lib/oauth/callback.js:112:16)
    at async Object.callback (/app/node_modules/next-auth/core/routes/callback.js:50:11)
    at async NextAuthHandler (/app/node_modules/next-auth/core/index.js:139:28)
    at async NextAuthNextHandler (/app/node_modules/next-auth/next/index.js:21:19)
    at async /app/node_modules/next-auth/next/index.js:57:32
    at async Object.apiResolver (/app/node_modules/next/dist/server/api-utils/node.js:184:9)
    at async NextNodeServer.runApi (/app/node_modules/next/dist/server/next-server.js:396:9) {
  name: 'OAuthCallbackError',
  code: undefined
}
```

Provider settings:
```
import NextAuth from 'next-auth';
import CognitoProvider from 'next-auth/providers/cognito';

export default NextAuth({
    providers: [
        CognitoProvider({
            clientId: process.env.COGNITO_CLIENT_ID,
            clientSecret: process.env.COGNITO_CLIENT_SECRET,
            issuer: process.env.COGNITO_ISSUER,
            idToken: true,
            checks: 'nonce',
        }),
    ],
    debug: true,
});
```

Container: node:14-alpine
next: 12.1.1
next-auth: ^4.3.1

**Reply by AntoineGerv** [reply]
Same problem here, 

next: 12.1.1
next-auth: ^4.3.1

**Reply by james-bjss** [reply]
Are you using the forked version of the code with this fix in it? The PR is not yet merged and has been readied for review in the next release.
See this comment from the maintainers here for a suggestion of test this out. Or you can build locally and drop the file into your node modules just for testing.
 https://github.com/nextauthjs/next-auth/pull/4100#issuecomment-1075692109

**Reply by EvgenKoshevoy** [reply]
Oh, i see. I thought it is already merged. I've used latest release version.

**Comment by jamesbluecrow** [comment]
Thanks for the effort @james-bjss . We are also affected by this issue, I'll try your forked version to see if it addresses the issue.

I'm mostly posting to add more interest on this fix being merged 😃

**Reply by jamesbluecrow** [reply]
> Thanks for the effort @james-bjss . We are also affected by this issue, I'll try your forked version to see if it addresses the issue.
> 
> I'm mostly posting to add more interest on this fix being merged 😃

I'm must be missing something, I did checkout your branch @james-bjss and run `yarn install` and `yarn build`. Then replaced the `next-auth` inside the `node_modules` in my application with the `packages/next-auth` from your branch but I still see the same error. Anything you think I'm doing wrong?

**Reply by hamidbjss** [reply]
@jamesbluecrow Use this patch instead and place into `patches/next-auth+4.3.3.patch`: https://gist.github.com/hamidbjss/b6408e48080f247edd22ec04a3b983e3

You'll need to install this package from npm [patch-package](https://www.npmjs.com/package/patch-package).

Finally, ensure you have the following script in your package.json scripts:
```
    "postinstall": "patch-package",
```

Next time you run `npm i` it will perform the postinstall step which will patch the next-auth package for you.

**Comment by AntoineGerv** [comment]
@lluia @balazsorban44 Can you help us ? PR is here : https://github.com/nextauthjs/next-auth/pull/4100
Thx

**Reply by james-bjss** [reply]
> @lluia @balazsorban44 Can you help us ? PR is here : #4100 Thx

Aaand it's merged! :tada: Thanks all.

**Comment by hamidbjss** [comment]
Created a patch for version 4.10.3 if it's of help to anyone [next-auth+4.10.3.patch](https://gist.github.com/hamidbjss/66f72a6f74784be3d32516c74c850d40#file-next-auth-4-10-3-patch).

**Comment by revmischa** [comment]
I have the latest package with this fix in it at `https://github.com/jetbridge/next-auth/raw/tmp/packages/next-auth/next-auth-4.10.3.tgz`

Correct me if I'm wrong but I believe you also need to add this to your provider config?
```ts
      checks: 'nonce',
```

**Reply by james-bjss** [reply]
Correct, there are other checks so you may want to append this to the existing ones. Namely the state and pkce options (can check the source for your given provider to see the defaults).
```ts
type ChecksType = "pkce" | "state" | "none" | "nonce"

**Reply by revmischa** [reply]
What checks should be used with cognito?

**Reply by james-bjss** [reply]
> What checks should be used with cognito?

pkce, state, nonce

**Comment by ronaldwh84** [comment]
Hi @james-bjss 

I am using the latest version 4.10.3 but it is not working. Do I need to perform some extra steps to apply this fix?

**Reply by james-bjss** [reply]
> Hi @james-bjss
> 
> I am using the latest version 4.10.3 but it is not working. Do I need to perform some extra steps to apply this fix?

The fix has been merged to main but there has been no new release since. 

You would need to apply the patch package as per convos above until the next nextauth release with the fix in it.

There is a flurry of chore activity on main at the moment from the maintainers and it looks like they may be prepping for a new release? So if you can wait it out the new package should be out eventually...

**Comment by ronaldwh84** [comment]
Ah ok noted that!

Thank you very much for the info, I will use the patch for now.

Cheers,
Ronaldwh

**Reply by james-bjss** [reply]
This is now released, though seems absent form the release notes. Can confirm its working for us without the patch on the latest release.
Maybe we can resolve this convo now?

**Reply by mjdye** [reply]
I'm currently on version 4.1.2 and still receiving this issue. Is there something in particular I'm missing?

```
CognitoProvider({
	clientId: '**',
	clientSecret: '**',
	issuer: '**',
	idToken: true,
        checks: 'nonce',
}),
```

The error is:
```
 providerId: 'cognito',
  message: 'nonce mismatch, expected undefined, got: someNonce
  ```

To be clear, I'm currently using the Cognito Logout URL to force Google to show the account selection screen again. This occurs when I swap accounts.

**Comment by z0ph** [comment]
I'm running the same behavior with:

```json
"next": "^13.0.5",
"next-auth": "^4.17.0",
```

Error:
![CleanShot 2022-11-29 at 14 29 57@2x](https://user-images.githubusercontent.com/20846187/204541772-9e5cea47-c247-4d8b-a7f6-a3524c25a2ac.png)


MyProvider:
```js
import NextAuth from "next-auth";
import CognitoProvider from "next-auth/providers/cognito";

export default NextAuth({
  providers: [
    // https://github.com/nextauthjs/next-auth/discussions/3551
    CognitoProvider({
      clientId: process.env.COGNITO_CLIENT_ID,
      clientSecret: process.env.COGNITO_CLIENT_SECRET,
      issuer: process.env.COGNITO_ISSUER,
      idToken: true,
      checks: 'nonce',
    }),
  ],
  callbacks: {
    // By default NextAuth will redirect to an url of the same hostname
    // this callback allows to change that behavior and specify a custom url
    redirect({ url }) {
      return url;
    },
    async session({ session, token }) {
      // Send properties to the client, like an access_token from a provider.
      session.customerID = token.sub;
      return session;
    },
  },

  secret: process.env.SECRET,
  debug: process.env.NODE_ENV === "development",
});

```

**Comment by akd-io** [comment]
Was able to fix the issue using the `checks: "nonce"` feature added i #4100.

My config:
```
CognitoProvider({
  clientId: process.env.COGNITO_CLIENT_ID,
  clientSecret: process.env.COGNITO_CLIENT_SECRET,
  issuer: process.env.COGNITO_ISSUER,
  checks: "nonce", // Necessary to fix a bug with Cognito's third-party providers. See the following issue and PR. Issue: https://github.com/nextauthjs/next-auth/discussions/3551 PR: https://github.com/nextauthjs/next-auth/pull/4100
}),
```

@leonp-s I spent quite a while searching through this thread to eventually find the right solution, maybe one of these comments linking to #4100 can be marked as the correct answer?

**Reply by akd-io** [reply]
Alright, I had it come back in the case where I have a NextAuth session, but not a Cognito session. I don't think my users will face this case though.

**Comment by half2me** [comment]
I'm having the same issue: https://github.com/nextauthjs/next-auth/issues/6865

**Reply by mortezaalizadeh** [reply]
I am having the same issue too, I am using version 4.22.1

**Reply by half2me** [reply]
I have `@auth/core` pinned to `0.5.1` and I apply this patch:
```
--- node_modules/@auth/core/lib/oauth/callback.js	2023-03-28 16:30:20
+++ node_modules/@auth/core/lib/oauth/callback.js	2023-04-17 14:11:07
@@ -76,7 +76,8 @@
     let profile = {};
     let tokens;
     if (provider.type === "oidc") {
-        const result = await o.processAuthorizationCodeOpenIDResponse(as, client, codeGrantResponse);
+        const result = await o.processAuthorizationCodeOpenIDResponse(as, client, codeGrantResponse, nonce?.value ??
+o.expectNoNonce);
         if (o.isOAuth2Error(result)) {
             console.log("error", result);
             throw new Error("TODO: Handle OIDC response body error");
```

That fixes things for me as long as I set `checks: ['pkce', 'nonce'],` for my Cognito provider settings.
Until this is properly fixed, that's the best I can do, but it does work quite well.

**Reply by mortezaalizadeh** [reply]
cheers @half2me  I can't really patch it, so decided to go with overriding the checks to nonce only and it worked like a charm

**Reply by half2me** [reply]
> cheers @half2me I can't really patch it, so decided to go with overriding the checks to nonce only and it worked like a charm

Keep it mind that will only work for normal username/password logins. If you try to use social login with cognito (login with facebook, google, etc.) it will give you an error.

**Comment by rvdende** [comment]
Should we perhaps add the fix as a default in https://github.com/nextauthjs/next-auth/blob/main/packages/core/src/providers/cognito.ts#LL91C10-L91C10 ?

```ts
options: { checks: "nonce", ...options }
```

**Reply by half2me** [reply]
it doesn't fix the problem though. If you use "social login" with cognito, it still gives you the same error

**Reply by rvdende** [reply]
works on my side for an azure active directory login...

**Comment by PorridgeBear** [comment]
This same exact nonce mismatch error with `expected undefined` but getting a value from Cognito is happening in NextAuth v5 beta, and no permutation of `checks` fixes the issue having tried `["pkce", "state"]` to exclude `nonce`, or `["pkce", "state", "nonce"]` to include it. The callback from Cognito sends a nonce, Auth.js v5 is making the comparison no matter the provider config, and the user get the Server Configuration error page. It is then possible to simply tweak the URL back to login and it works second time. I know NextAuth v4 does not have this issue anymore, so will need to revert until solved in v5.
